tests/framework/pvc_utils.go

```diff
@@ -0,0 +1,192 @@
+	var pod *k8sv1.Pod
+	podDef := newExecutorPodWithPVC(podName, "ls -1 /pvc | wc -l", pvc)
+	err := wait.PollImmediate(2*time.Second, DefaultTimeout, func() (bool, error) {
+		var err error
+		pod, err = f.K8sClient.CoreV1().Pods(f.Namespace.GetName()).Create(podDef)
+		if err != nil {
+			return false, nil
+		}
+		return true, nil
+	})
+	if err != nil {
+		Fail("Unable to create Pod: " + podName + ", error: " + err.Error())
+	}
+	return pod
+}
+
+// Delete the passed in Pod from the passed in Namespace
+func (f *Framework) DeletePodFromNamespace(pod *k8sv1.Pod, namespace *k8sv1.Namespace) {
+	err := wait.PollImmediate(2*time.Second, DefaultTimeout, func() (bool, error) {
+		err := f.K8sClient.CoreV1().Pods(namespace.GetName()).Delete(pod.GetName(), &metav1.DeleteOptions{})
+		if err != nil {
+			return false, nil
+		}
+		return true, nil
+	})
+	if err != nil {
+		Fail("Unable to delete Pod: " + pod.GetName() + ", error: " + err.Error())
+	}
+}
+
+// Delete the passed in Pod from the Framework primary namespace
+func (f *Framework) DeletePod(pod *k8sv1.Pod) {
+	f.DeletePodFromNamespace(pod, f.Namespace)
+}
+
+func newExecutorPodWithPVC(podName, cmd string, pvc *k8sv1.PersistentVolumeClaim) *k8sv1.Pod {
+	return &k8sv1.Pod{
+		ObjectMeta: metav1.ObjectMeta{Name: podName},
+		Spec: k8sv1.PodSpec{
+			RestartPolicy: k8sv1.RestartPolicyNever,
+			Containers: []k8sv1.Container{
+				{
+					Name:    "runner",
+					Image:   "fedora:28",
+					Command: []string{"/bin/sh", "-c", "sleep 5; echo I am an executor pod;"},
+					VolumeMounts: []k8sv1.VolumeMount{
+						{
+							Name:      pvc.GetName(),
+							MountPath: defaultPvcMountPath,
+						},
+					},
+				},
+			},
+			Volumes: []k8sv1.Volume{
+				{
+					Name: pvc.GetName(),
+					VolumeSource: k8sv1.VolumeSource{
+						PersistentVolumeClaim: &k8sv1.PersistentVolumeClaimVolumeSource{
+							ClaimName: pvc.GetName(),
+						},
+					},
+				},
+			},
+		},
+	}
+}
+
+func (f *Framework) waitTimeoutForPodReady(podName string, timeout time.Duration) error {
+	return wait.PollImmediate(2*time.Second, timeout, f.podRunning(podName, f.Namespace.GetName()))
+}
+
+func (f *Framework) podRunning(podName, namespace string) wait.ConditionFunc {
+	return func() (bool, error) {
+		pod, err := f.K8sClient.CoreV1().Pods(namespace).Get(podName, metav1.GetOptions{})
+		if err != nil {
+			return false, err
+		}
+		switch pod.Status.Phase {
+		case k8sv1.PodRunning:
+			return true, nil
+		case k8sv1.PodFailed, k8sv1.PodSucceeded:
+			return false, nil
+		}
+		return false, nil
+	}
+}
+
+// Verify a particular PVC is empty, returns true if it is, false otherwise.
+func (f *Framework) VerifyPVCIsEmpty(pvc *k8sv1.PersistentVolumeClaim) bool {
+	executorPod := f.CreateExecutorPodWithPVC("verify-pvc-empty", pvc)
+	err := f.waitTimeoutForPodReady(executorPod.GetName(), DefaultTimeout)
+	Expect(err).To(BeNil())
```

Comment:
My vote is that all framework funcs and methods return err rather than calling `Expect`, `Fail` or panic. I think this is the best practice for a lower-level, common framework pkg.

Reply:
This is a good question, and I am not sure I have a good answer. Where do we draw the line on what is a utility function, and what is a framework function. This one in particular is a good one to ask that question. Its a higher level function that calls a bunch of lower level ones, so I would lean more towards a utility function. But the other ones are all single purpose functions, but they will now all take namespace as a parameter.

Maybe put them all in a utility package, and put matching functions without the namespace in the framework, and delegate the work to the utility functions with the namspace filled in.

In short, I don't know, maybe we need to talk about that.

Reply:
In my view, all functions needed to set up the running cluster so that general e2e tests can be reliably and repeatedly run should be in the framework pkg. This may eventually consist of several files but now there's just one.  So this includes namespaces, service-acct, cluster roles, services, maybe an executor pod, possibly a general purpose, common storage class...
Helper functions used to manage more specialized resource such as importer pod, pvc, pv, ... should live in a _utils_ pkg. The framework pkg never calls Expect or Fail or panic -- always returns err. (Though the Before and AfterEach's in framework call Expect now). The functions in a _utils_ pkg may call Expect with the assumption they are **always** called in a ginkgo Describe, Context or It block. What do you think?